## To predict if a customer will repay loan amount or not using Decision Tree algorithm

## Import Libraries

In [1]:
import numpy as np
import pandas as pd

In [2]:
data=pd.read_csv('C:\python\Machine learning\Datasets\Decision_Tree_ Dataset.csv',sep=',',header=0)
data.head()

Initial payment  Last payment  Credit Score  House Number    sum Result
0              201         10018           250          3046  13515    yes
1              205         10016           395          3044  13660    yes
2              257         10129           109          3251  13746    yes
3              246         10064           324          3137  13771    yes
4              117         10115           496          3094  13822    yes

In [3]:
len(data)

1000

In [4]:
data.shape

(1000, 6)

## Data Cleaning

In [5]:
data=data.drop(('sum'),axis=1)

In [6]:
data.head()

Initial payment  Last payment  Credit Score  House Number Result
0              201         10018           250          3046    yes
1              205         10016           395          3044    yes
2              257         10129           109          3251    yes
3              246         10064           324          3137    yes
4              117         10115           496          3094    yes

In [7]:
data.isnull().sum()

Initial payment    0
Last payment       0
Credit Score       0
House Number       0
Result             0
dtype: int64

## Divide the data into dependent and independent variable

In [8]:
X=data.iloc[:,:-1].values
X

array([[  201, 10018,   250,  3046],
       [  205, 10016,   395,  3044],
       [  257, 10129,   109,  3251],
       ...,
       [  316, 14872,   613,  4760],
       [  305, 14926,   897,  4572],
       [  168, 14798,   834,  4937]], dtype=int64)

In [9]:
Y=data.iloc[:,-1].values
Y

array(['yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes',
       'yes', 'yes',

## Split the data into Training and Testing 

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.3,random_state=0)

## Fit the model to a data

In [12]:
from sklearn.tree import DecisionTreeClassifier

In [13]:
model=DecisionTreeClassifier(criterion='entropy',random_state=0,max_depth=3,min_samples_leaf=5)

In [14]:
model.fit(X_train,Y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=3, min_samples_leaf=5,
                       random_state=0)

## Prediction

In [15]:
Y_pred=model.predict(X_test)
Y_pred

array(['No', 'No', 'yes', 'yes', 'No', 'No', 'yes', 'yes', 'yes', 'No',
       'No', 'No', 'No', 'yes', 'No', 'yes', 'No', 'yes', 'yes', 'yes',
       'yes', 'yes', 'yes', 'yes', 'No', 'No', 'yes', 'yes', 'No', 'yes',
       'No', 'No', 'No', 'No', 'yes', 'No', 'No', 'yes', 'No', 'yes',
       'yes', 'No', 'No', 'No', 'yes', 'No', 'No', 'yes', 'yes', 'No',
       'yes', 'yes', 'yes', 'yes', 'No', 'No', 'yes', 'yes', 'No', 'yes',
       'No', 'No', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'No',
       'No', 'yes', 'No', 'yes', 'No', 'yes', 'yes', 'No', 'yes', 'No',
       'yes', 'yes', 'yes', 'No', 'No', 'No', 'yes', 'No', 'yes', 'No',
       'yes', 'No', 'No', 'yes', 'No', 'No', 'No', 'yes', 'yes', 'No',
       'No', 'No', 'yes', 'No', 'yes', 'No', 'No', 'No', 'No', 'yes',
       'yes', 'No', 'No', 'No', 'yes', 'yes', 'yes', 'yes', 'No', 'No',
       'No', 'No', 'yes', 'No', 'yes', 'No', 'yes', 'yes', 'No', 'yes',
       'yes', 'No', 'yes', 'No', 'No', 'No', 'yes', 'yes', 'yes'

## Accuracy and Confusion Matrix

In [16]:
from sklearn.metrics import confusion_matrix,accuracy_score

In [17]:
print('Accuracy is',accuracy_score(Y_test,Y_pred)*100)

Accuracy is 92.0


In [18]:
print("Confusion matrix is",confusion_matrix(Y_test,Y_pred))

Confusion matrix is [[151  16]
 [  8 125]]


So, we have created a model that uses decision tree algorithm to predict whether a customer will repay the loan or not with accuracy score of the model is 92%. The bank can use this model to decide whether it should approve loan request from a customer or not.